yarn-project/sequencer-client/src/sequencer/sequencer.ts

```diff
@@ -285,38 +259,204 @@ export class Sequencer {
-    [Attributes.BLOCK_NUMBER]: newGlobalVariables.blockNumber.toNumber(),
-  }))
+  async canProposeBlock(
+    historicalHeader?: Header,
+    globalVariables?: GlobalVariables,
+    tipArchive?: Buffer,
+  ): Promise<bigint> {
+    // @note  In order to be able to propose a block, a few conditions must be met:
+    //        - We must be caught up to the pending chain
+    //          - The tip archive must match the one from the L1
+    //          - The block number should match the one from the L1
+    //        - If we have built a block, the block number must match the next pending block
+    //        - There cannot already be a block for the slot
+    //        - If we are NOT in devnet, then the active slot must match the slot number of the block
+    //        - The proposer must either be free for all or specifically us.
+    //
+    //        Note that the ordering of these checks are NOT optimised for performance, but to resemble the ordering
+    //        that is used in the Rollup contract:
+    //        - _validateHeaderForSubmissionBase
+    //        - _validateHeaderForSubmissionSequencerSelection
+    //
+    //        Also, we are logging debug messages for checks that fail to make it easier to debug, as we are usually
+    //        catching the errors.
+
+    const [submitter, slot, pendingBlockNumber, archive] = await this.publisher.getMetadataForSlotAtNextEthBlock();
+
+    if (await this.publisher.willSimulationFail(slot)) {
+      // @note  See comment in willSimulationFail for more information
+      const msg = `Simulation will fail for slot ${slot}`;
+      this.log.debug(msg);
+      throw new Error(msg);
+    }
+
+    // If our tip of the chain is different from the tip on L1, we should not propose a block
+    // @note  This will change along with the data publication changes.
+    if (tipArchive && !archive.equals(tipArchive)) {
+      const msg = `Tip archive does not match the one from the L1`;
+      this.log.debug(msg);
+      throw new Error(msg);
+    }
+
+    // Make sure I'm caught up to the pending chain
+    if (
+      pendingBlockNumber > 0 &&
+      (historicalHeader == undefined || historicalHeader.globalVariables.blockNumber.toBigInt() != pendingBlockNumber)
+    ) {
+      const msg = `Not caught up to pending block ${pendingBlockNumber}`;
+      this.log.debug(msg);
+      throw new Error(msg);
+    }
+
+    // If I have constructed a block, make sure that the block number matches the next pending block number
+    if (globalVariables) {
+      if (globalVariables.blockNumber.toBigInt() !== pendingBlockNumber + 1n) {
+        const msg = `Block number mismatch. Expected ${
+          pendingBlockNumber + 1n
+        } but got ${globalVariables.blockNumber.toBigInt()}`;
+        this.log.debug(msg);
+        throw new Error(msg);
+      }
+
+      const currentBlockNumber = await this.l2BlockSource.getBlockNumber();
```

Comment:
How could this happen if we have ensured that L1 is expecting this block number?